### ISSUE TITLE
Lint for iterator.chain(None)

### ISSUE THREAD

**Issue by enterprisey** [rust-lang/rust-clippy#6214]
### What it does

Catches `.chain(None)` being called on an iterator, in which case the author probably meant `.chain(std::iter::once(None))`.

### Categories (optional)

- Kind: correctness, probably - `.chain(None)` is a no-op!

*What is the advantage of the recommended code over the original code*

Actually does what the user probably intended to.

### Drawbacks

Very, very specific lint. Maybe too specific.

### Example

```rust
let iterator = vec![Some(1)].chain(None);
```

Could be written as:

```rust
let iterator = vec![Some(1)].chain(std::iter::once(None));
```

**Comment by llogiq** [comment]
But sometimes folks might want to have types line up, and have multiple code paths where one path chains a `Some(value)` at the end. So the correct way to create a value off the same type without appended value would be `.chain(None)`.

**Comment by enterprisey** [comment]
Never thought of that! I guess there's a use-case for most everything.
